### ISSUE TITLE
Rocket State

### ISSUE THREAD

**Issue by Alakazam03** [SergioBenitez/Rocket#1147]
```
struct HitCount {
    sto : db::Db,
}
fn main() {
    rocket::ignite()
        .manage(HitCount{ sto : db::Db::new("test".to_string(), true)})
        .mount("/vid", routes![vid_generation::generate_splits, vid_generation::recover_secret])
        .mount("/add", routes![smt_number::add_customer])
        .launch();
}
```
i am trying to initiate a db connection for storing merkle tree. I want to use a global merkle tree using state.
I am getting error error: `std::rc::Rc<std::boxed::Box<(dyn rusty_leveldb::cmp::Cmp + 'static)>>` cannot be sent between threads safely
label: `std::rc::Rc<std::boxed::Box<(dyn rusty_leveldb::cmp::Cmp + 'static)>>` cannot be sent between threads safely

I have been struggling alot and have tried using lazy_static, Once_cell! Please help @SergioBenitez

**Comment by DanSnow** [comment]
You may need to wrap your `Db` with a `Mutex` since it can not share between thread.

https://github.com/SergioBenitez/Rocket/blob/master/examples/raw_sqlite/src/main.rs

**Comment by jebrosen** [comment]
Types in manged state must be thread-safe (that is, `Send` + `Sync`). `std::rc::Rc` is not `Send`, which will be a problem.

What type is `Db` in your example?

**Comment by Alakazam03** [comment]
> Types in manged state must be thread-safe (that is, `Send` + `Sync`). `std::rc::Rc` is not `Send`, which will be a problem.
> 
> What type is `Db` in your example?
It is rustylevel_db. I am using a package from https://github.com/arnaucube/merkletree-rs.
Thanks for the help.

**Comment by jebrosen** [comment]
> @jebrosen Please can you reply?

For future reference edits are silent, so I did not know you pinged until now.

---

I looked closer at `rusty_leveldb::DB` and it is not `Send` because it uses `Rc`. I don't know if making it `Send` would be easy or difficult to do.

Finding another implementation of the merkle tree and/or leveldb is probably the easiest option, if you haven't yet written a lot of code using it. Another workaround could be to have a dedicated thread for reading and writing the DB, and use channels to send update commands to that thread.
